src/IO/ADIOS/CommonADIOS1IOHandler.cpp

```diff
@@ -276,8 +311,8 @@ CommonADIOS1IOHandlerImpl::flush_attribute(int64_t group, std::string const& nam
         }
         case DT::BOOL:
         {
-            auto ptr = reinterpret_cast< unsigned char* >(values.get());
-            *ptr = static_cast< unsigned char >(att.get< bool >());
+            auto ptr = reinterpret_cast< bool* >(values.get());
+            *ptr = static_cast< bool >(att.get< bool >());
```

Comment:
I am unsure about this.
As far as I can remember, ADIOS does not offer support for native `bool` due to their C origins.
Which is why there was a [workaround](https://github.com/openPMD/openPMD-api/pull/126/commits/ccd9351709a6438583bb36a73140ff4d17f0477c) borrowed from [PIConGPU](https://github.com/ComputationalRadiationPhysics/picongpu/pull/1756).

Reply:
oh yes, let me revert this. I thought about this as well and was reckless.

Reply:
I just realized I intentionally changed this while updating the depending PRs :) This is the read part, of course we want an actual `bool` back here. Will change it therein.

Reply:
> This is the read part

Lines [314-315](https://github.com/ax3l/openPMD-api/blob/159d8f4f7effd0f17e5af4ee236e491c6cebb3ed/src/IO/ADIOS/CommonADIOS1IOHandler.cpp#L314-L315) are actually part of [```CommonADIOS1IOHandler::flushAttribute()```](https://github.com/ax3l/openPMD-api/blob/159d8f4f7effd0f17e5af4ee236e491c6cebb3ed/src/IO/ADIOS/CommonADIOS1IOHandler.cpp#L40-L42). So unless you count the transfer from in-memory to on-disk a _read_, then no. Then this is not the read-API. Changining this to `bool` will break ADIOS (for the stated reason).

The place where something along your lines can be implemented is in lines [1022-1025](https://github.com/ax3l/openPMD-api/blob/159d8f4f7effd0f17e5af4ee236e491c6cebb3ed/src/IO/ADIOS/CommonADIOS1IOHandler.cpp#L1022-L1025) of [```CommonADIOS1IOHandler::readAttribute()```](https://github.com/ax3l/openPMD-api/blob/159d8f4f7effd0f17e5af4ee236e491c6cebb3ed/src/IO/ADIOS/CommonADIOS1IOHandler.cpp#L865-L868). The problem here is different, though. The Attribute is stored as `uint8_t and we can not determine whether it is a true `uint8_t` or a workaround for bool (casting for values `0` and `1` will lead to even more havoc).

This cast is better suited in your proposal #335 for unifying `Attribute` and `Variant`. This ADIOS1 corner-case would then be appropriately handled by `getCast`.

Reply:
This is not entirely invalid, though.
```cpp
case DT::BOOL:
{
    /* values is a void-pointer that is given to ADIOS
     * to support bool writing, this will have to be allocated as uint8_t */
    auto ptr = reinterpret_cast< unsigned char* >(values.get());
    /* the current workaround is heaved into the frontend
     * (i.e. the user has to pass uint8_t)
     * the backend can do the appropriate widening from bool to uint8_t */
    *ptr = static_cast< unsigned char >(att.get< bool >());
    break;
}
```

Reply:
Thanks, that's important information I missed!

Just to make things worse: As I just remembered from the special case of `std::vector< bool >` in the stdlibs is that a bool is not necessarily the same size as an `uint8_t` ;)
https://stackoverflow.com/questions/4897844/is-sizeofbool-defined

Of course I don't know any implementation where this is not a Byte, but who knows ;-)

Reply:
> a bool is not necessarily the same size as an uint8_t

That should not be a problem.
```cpp
/* frontend */
Series s = Series("bool_test.bp", AccessType::CREATE);
bool attribute_value = true;
s.setAttribute("bool", attribute_value); /* schedules a IOTask with a bool */
s.flush();

/* ADIOS1 backend steps in pseudo-code 
 * 1. schedule attribute write, not modifying any parameters yet
 * 2. in flushAttribute(), allocate a uint8_t behind a void* to give to ADIOS
 * 3. retrieve the bool as a bool (!) from the Attribute
 * 4. static_cast the bool to uint8_t, store it in the void*
 * 5. give  void* to ADIOS
 * 6. ???
 * 7. profit */
```

In this case, it won't even matter what `sizeof(bool)` is, as `0`/`1` can *always* be stored in `uint8_t`.